### PR TITLE
[WIP] cache/remote: drop dos2unix MD5 by default

### LINKED ISSUE #4658: drop dos2unix behaviour
# DVC CRLF issue

I've prepared a sample repository with the steps to reproduce this issue. 
https://github.com/MetalBlueberry/dvc-crlf

```txt
DVC version: 1.7.2 (pip)
---------------------------------
Platform: Python 3.6.9 on Linux-5.3.0-28-generic-x86_64-with-Ubuntu-18.04-bionic
Supports: All remotes
Cache types: hardlink, symlink
Cache directory: ext4 on /dev/nvme0n1p5
Workspace directory: ext4 on /dev/nvme0n1p5
Repo: dvc, git
```

> Cache is committed to the repo so it can be viewed

## Reproduce

- init git and dvc
- create a script.sh with CRLF line endings
- dvc add script.sh
- change CRLF ending to LF
- try to add it again, (doesn't detect changes)

## Obtained

script.sh is stored with CRLF in cache, so when you run `dvc checkout` it always comes back to CRLF line endings even if you modify it locally.

## Expected

script.sh should keep the original CRLF line endings or at least store the cache content always with LF endings.

## Consequences

- CRLF and LF cannot be modified after first commit
- md5 file name doesn't match the file content hash

Thank you

First reported in https://github.com/iterative/dvc.org/issues/1469#issuecomment-702129969

### PR DESCRIPTION
**Pull request by pmrowla**
* [ ] ❗ I have followed the [Contributing to DVC](https://dvc.org/doc/user-guide/contributing/core) checklist.

* [ ] 📖 If this PR requires [documentation](https://dvc.org/doc) updates, I have created a separate PR (or issue, at least) in [dvc.org](https://github.com/iterative/dvc.org) and linked it here.

Thank you for the contribution - we'll try to review it as soon as possible. 🙏

Will close #4658 (replaces https://github.com/iterative/dvc/pull/5337)

### CODE REVIEW

File: dvc/utils/__init__.py [hunk 1]
```diff
@@ -41,8 +41,8 @@ def _fobj_md5(fobj, hash_md5, binary, progress_func=None):
             progress_func(len(data))
 
 
-def file_md5(fname, tree=None):
-    """ get the (md5 hexdigest, md5 digest) of a file """
+def file_md5(fname, tree=None, enable_d2u=False):
```
Comment (efiop) [inline dvc/utils/__init__.py:44]: Let's create two functions:
```
file_md5
file_dos2unixmd5
```
or
```
get_md5
get_dos2unixmd5
```
just so we don't get tangeled in it ourselves :D

File: dvc/objects.py [hunk 2]
```diff
@@ -19,14 +19,17 @@ class ObjectFormatError(ObjectError):
...
         return tree.get_hash(path_info, **kwargs)
 
     if name == "md5":
         from dvc.hash_info import HashInfo
+        from dvc.odb.versions import ODB_VERSION
         from dvc.utils import file_md5
 
+        if odb.version == ODB_VERSION.V1:
+            kwargs["enable_d2u"] = True
         return HashInfo("md5", file_md5(path_info, tree, **kwargs))
```
Comment (efiop) [inline dvc/objects.py:33]: stuff under `if name == "md5"` is defunct, just added to clarify the intentions.

We should actually do `tree.get_hash(path_info, name, **kwargs)` and so then, for example, LocalTree.get_hash will see that we ask it about `md5` and will use `file_md5` and if it sees `dos2unixmd5` - `file_dos2unixmd5`

**Comment by efiop** [comment]
Sorry for the noise, I understand that this is a WIP and needs to be adapted.

File: dvc/utils/__init__.py (hunk 1 above)
Comment (efiop) [reply dvc/utils/__init__.py:44]: Though a flag is good too. Nevermind :slightly_smiling_face:

File: dvc/remote/base.py [hunk 3]
```diff
@@ -90,6 +92,15 @@ def __repr__(self):
...
 
+    @property
+    def version(self):
+        from dvc.odb.versions import LATEST_VERSION
+        from dvc.parsing.versions import SCHEMA_KWD
+
+        if self._config:
+            return self._config.get(SCHEMA_KWD, LATEST_VERSION)
+        return LATEST_VERSION
+
```
Comment (efiop) [inline dvc/remote/base.py:103]: Could we just use `self.cache.version`?

File: dvc/hash_info.py [hunk 4]
```diff
@@ -7,6 +8,17 @@
 DirInfo = Dict[str, str]
 
 
+class HashName(str, enum.Enum):
+    MD5 = "md5"  # Raw MD5
+    MD5_D2U = "md5-d2u"  # DVC dos2unix MD5
```
Comment (pmrowla) [inline dvc/hash_info.py:13]: As it's written right now, for users who have explicitly opted into using dos2unix, we will write a config file containing version `1.0`, and dvc.lock files will now contain hashes like

```
md5-d2u: abc123
```

This seems more "correct" to me versus continuing to write lock files with `md5: abc123`, but the downside would be that anyone on DVC < 2.0 won't be able to read lock files with `md5-d2u` hash names.

File: dvc/odb/config.py [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+
+    logger.debug("Writing ODB config '%s'", path_info)
+    if not fs.exists(path_info.parent):
+        fs.makedirs(path_info.parent)
+
+    with NamedTemporaryFile() as tmp:
+        with modify_yaml(tmp.name) as data:
+            data.update(config)
+        tmp.seek(0)
+        fs.upload_fobj(tmp, path_info)
```
Comment (pmrowla) [inline dvc/odb/config.py:52]: This seems like a good use case for `MemoryFileSystem`, but for some reason the fsspec memory implementation of `fs.open()` does not play nicely with the edit-in place `modify_yaml` functionality and I didn't think it was worth digging into the issue at this point.

**Comment by pmrowla** [comment]
Still need to investigate why the last 2 tests pass for me locally but fail in CI

**Pull request closed without merging**